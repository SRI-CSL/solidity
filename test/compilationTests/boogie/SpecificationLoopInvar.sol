pragma solidity ^0.4.23;

contract SpecificationLoopInvar {
    function correct() public pure {
        int i = 0;
        int x = 0;
        /** @notice invariant i == x */
        while (i < 10) {
            i++;
            x++;
        }
        assert(x == i);
    }
    function cannotProveWithoutInvariant() public pure {
        int i = 0;
        int x = 0;
        while (i < 10) {
            i++;
            x++;
        }
        assert(x == i); // Cannot prove assertion without invariant
    }
    function doesNotHoldOnEntry() public pure {
        int i = 0;
        int x = 1; // Invariant does not hold on entry
        /** @notice invariant i == x */
        while (i < 10) {
            i++;
            x++;
        }
        assert(x == i);
    }
    function isNotMaintainedByLoop() public pure {
        int i = 0;
        int x = 0;
        /** @notice invariant i == x */
        while (i < 10) {
            i++; // Invariant not maintained by loop
        }
        assert(x == i);
    }
}