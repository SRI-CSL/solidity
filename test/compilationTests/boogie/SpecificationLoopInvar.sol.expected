^Boogie program verifier finished with 2 verified, 3 errors$